File: 438-find-all-anagrams-in-a-string/438-find-all-anagrams-in-a-string.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        n, k, i, j = len(s), len(p), 0, 0
        ans, mp = [], {}
        for c in p:
            if c in mp:
                mp[c] += 1
            else:
                mp[c] = 1
        chars = len(mp)
        
        while j < n:
            if s[j] in mp:
                mp[s[j]] -= 1
                if mp[s[j]] == 0: 
                    chars -= 1
            if j - i + 1 < k:
                j += 1
            elif j - i + 1 == k:
                if chars == 0:
                    ans.append(i)
                if s[i] in mp:
                    mp[s[i]] += 1
                    if mp[s[i]] == 1:
                        chars += 1
                i, j = i + 1, j + 1
                
        return ans

File: 438-find-all-anagrams-in-a-string/438-find-all-anagrams-in-a-string.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> ans = new ArrayList<>();
        int n = s.length(), k = p.length(), i = 0, j = 0;
        HashMap<Character, Integer> mp = new HashMap<>();
        for(int idx = 0; idx < p.length(); ++idx) {
            if(mp.containsKey(p.charAt(idx))) 
                mp.put(p.charAt(idx), mp.get(p.charAt(idx)) + 1);
            else
                mp.put(p.charAt(idx), 1);
        }
        
        int chars = mp.size();
        
        while(j < n) {
            if(mp.containsKey(s.charAt(j))) {
                mp.put(s.charAt(j), mp.get(s.charAt(j)) - 1);
                if(mp.get(s.charAt(j)) == 0) chars--;
            }
            if(j - i + 1 < k) j++;
            else if(j - i + 1 == k) {
                if(chars == 0)
                    ans.add(i);
                if(mp.containsKey(s.charAt(i))) {
                    mp.put(s.charAt(i), mp.get(s.charAt(i)) + 1);
                    if(mp.get(s.charAt(i)) == 1) chars++;
                }
                i++;
                j++;
            }
        }
        return ans;
    }
}